// ottoman/test.js
var ottoman = require('./lib/ottoman');
var util = require('util');
var couchbase = require('couchbase');

var cluster = new couchbase.Mock.Cluster('couchbase://127.0.0.1');
var bucket = cluster.openBucket('default');
ottoman.bucket = bucket;

var UserMdl = ottoman.model('User', {
  name: 'string'
});

var user = new UserMdl({
  name: 'Brett Lawson'
});

console.log(user);
console.log(user.id());

var userx = UserMdl.ref(user.id());
console.log(userx);
console.log(userx.id());

return;

function PhoneValidator(val) {
  var phoneno = /^\(?([0-9]{3})\)?[-. ]?([0-9]{3})[-. ]?([0-9]{4})$/;
  if(!val.match(phoneno)) {
    throw new Error('Phone number is invalid.');
  }
}

var UsersMdl = ottoman.model('Users', {
  phone: {
    type: 'string',
    validator: PhoneValidator
  }
});

var user = new UsersMdl();
user.phone = '506-871-8108';
user.save(function(err) {
  console.log('ERR', err);
});


return;



console.log(UsersMdl.schema.fields);

var users = new UsersMdl();
users.mylist.push(new UserMdl({name:'Brett'}));
users.mylist.push(new UserMdl({name:'Joe'}));
users.mylist.push(new UserMdl({name:'John'}));
console.log(users.toJSON());

var usersCoo = ottoman.toCoo(users);
var usersX = ottoman.fromCoo(usersCoo);

console.log(usersX);

/*
UserMdl.create({name:'Brett'}, function(err, ux) {
  console.log(err, ux);

  UserMdl.create({name: 'Joe'}, function(err, uy) {
    console.log(err, uy);



    process.exit(0);
  });
});
*/

return;

var NoteMdl = ottoman.model('Note', {
  user: { ref: 'User' },
  msg: 'string'
});

var UserMdl = ottoman.model('User', {
  name: 'string',
  note: NoteMdl
});

var user = new UserMdl();
var note = new NoteMdl();

note.user = user;
note.msg = 'Hey there!';

user.name = 'Brett';

user.save(function(err) {
  if (err) console.error(err);

  note.save(function(err) {
    if (err) console.error(err);

    console.log(user);
    console.log(note);

    NoteMdl.getById(note.id(), function(err, userX) {
      if (err) console.error(err);

      console.log('LOADED', userX);
      userX.load('user', function(err) {
        if (err) console.error(err);

        console.log('LOADED2', userX);
      })
    });
  });
});

return;

var TestMdl = ottoman.model('Test', {
  name: 'string',
  company: {
    type: 'string',
    default: 'unset'
  },
  when: 'Mixed',
  idx: 'integer',
  user: {
    name: 'string',
    _id: {
      type: 'string',
      auto: 'uuid',
      readonly: true
    },
    creator: {
      name: 'string',
      _id: 'string'
    }
  }
}, {
  id: 'user._id'
});

var x = new TestMdl();
x.user.name = 'Brett19';
var xJson = x.toJSON();

console.log(x);
console.log(xJson);

var y = ottoman.fromCoo(xJson);
console.log(y);


return;

var attemptSchema = ottoman.model('LoginAttempt', {
  ip: { type: 'string', default: '' },
  user: { type: 'string', default: '' },
  time: { type: 'Date', default: Date.now }
});

ottoman.models.LoginAttempt.create({
  ip: 'lol',
  user: 'Brett19'
}, function(err) {
  console.log(err);
  bucket.dump();
});

return;

exports = module.exports = function(app, mongoose) {
  var attemptSchema = new mongoose.Schema({
    ip: { type: String, default: '' },
    user: { type: String, default: '' },
    time: { type: Date, default: Date.now, expires: app.config.loginAttempts.logExpiration }
  });
  attemptSchema.index({ ip: 1 });
  attemptSchema.index({ user: 1 });
  attemptSchema.set('autoIndex', (app.get('env') === 'development'));
  app.db.model('LoginAttempt', attemptSchema);
};

return;

var TestMdl = ottoman.model('Test', {
  name: 'string',
  company: {
    type: 'string',
    default: 'unset'
  },
  when: 'Mixed',
  idx: 'integer',
  user: {
    name: 'string',
    _id: {
      type: 'string',
      auto: 'uuid',
      readonly: true
    },
    creator: {
      name: 'string',
      _id: 'string'
    }
  }
}, {
  id: 'user._id',
  index: {
    findByCompany: {
      type: 'view',
      by: 'company'
    },
    getByIdx: {
      type: 'refdoc',
      by: 'idx'
    }
  },
  queries: {
    topPosts: {
      of: 'Post',
      by: 'creator'
    }
  }
});

var PostMdl = ottoman.model('Post', {
  creator: { type: TestMdl, ref: true },
  company: 'string'
});

//console.log(util.inspect(TestMdl, {depth:100}));
var x = new TestMdl();
x.idx = 14;
x.when = new Date();
x.company = 'Couchbase';
console.log(util.inspect(x, {depth:100}));
console.log(util.inspect(x.toJSON(), {depth:100}));

var y = new TestMdl();
y.idx = 20;
y.company = 'VMware';

var z = new PostMdl();
z.creator = y;
z.company = 'Couchbase';

z.save(function(err) {
  console.log('SAVED z', err);

  y.save(function(err) {
    console.log('SAVED y', err);

    x.save(function (err) {
      console.log('SAVED1', err);
      bucket.dump();

      TestMdl.getByIdx(14, function (err, resObjs) {
        console.log('GET', err, res);

        var res = resObjs[0];
        res.idx = 12;
        res.save(function (err) {
          console.log('SAVED2', err);
          bucket.dump();

          console.log(res.toJSON());
          ottoman.ensureIndices(function () {
            console.log('Ensured Indexes');
            bucket.dump();


            TestMdl.findByCompany('Couchbase', function(err, res) {
              console.log('FOUND', err, res);
              console.log(z.toJSON());

              y.topPosts(function(err, res) {
                console.log('TOP POSTS', err, res);
              });
            });

          });
        });
      });

      /*
       var xRef = TestMdl.ref(x.id());
       console.log(xRef);

       xRef.load(function(err) {
       console.log('LOADED', err);
       console.log(xRef);

       TestMdl.getById(x.id(), function(err, xFound) {
       console.log('LOADED', err);
       console.log(xFound);
       });
       });
       */
    });
  });
});

return;
/*
try {
  x._id = 'TEST-FAIL';
} catch (e) {
  console.log('EXCEPT: ', e);
}
console.log(x._id);
*/

console.log(JSON.stringify(x));
console.log(x.id());

console.log('x:', x.loaded(), x);
var xCoo = ottoman.toCoo(x);
console.log('y:', xCoo);
var xObj = ottoman.fromCoo(xCoo);
console.log('z:', xObj.loaded(), xObj);



/*
console.log('Registered Models:');
for (var i in ottoman.models) {
  console.log('  - ' + i);
}
*/
